docs/extensibility.md

```diff
@@ -0,0 +1,122 @@
+
+## Add-on Extensions
+
+In a general case, implementation of an add-on is a class which implements the `ClusterAddOn` interface.
+
+```typescript
+export declare interface ClusterAddOn { 
+    id? : string;
+    deploy(clusterInfo: types.ClusterInfo): Promise<Construct> | void;
+}
+```
+
+**Note**: The add-on implementation can optionally supply the `id` attribute if the target add-on can be added to a blueprint more than once. 
+
+Implementation of the add-on is expected to be an exported class that implements the interface and supplies the implementation of the `deploy` method. In order for the add-on to receive the deployment contextual information about the provisioned cluster, region, resource providers and/or other add-ons, the `deploy` method takes the `ClusterInfo` parameter, which represents a structure defined in the SPI (service provider interface) contracts. The API for the cluster info structure is stable and provides access to the provisioned EKS cluster, scheduled add-ons (that have not been installed yet but are part of the blueprint) or provisioned add-ons and other contexts.
+
+### Post Deployment Hooks
+
+In certain cases, add-on provisioning may require logic to be executed after provisioning of the add-ons (and teams) is complete. For such cases, add-on can optionally implement `ClusterPostDeploy` interface.
+
+```typescript
+/**
+ * Optional interface to allow cluster bootstrapping after provisioning of add-ons and teams is complete.
+ * Can be leveraged to bootstrap workloads, perform cluster checks. 
+ * ClusterAddOn implementation may implement this interface in order to get post deployment hook point.
+ */
+export declare interface ClusterPostDeploy {
+    postDeploy(clusterInfo: types.ClusterInfo, teams: Team[]): void;
+}
+```
+
+This capability is leveraged for example in ArgoCD add-on to bootstrap workloads after all add-ons finished provisioning. Note, in this case unlike the standard `deploy` method implementation, the add-on also gets access to the provisioned teams. 
+
+### Add-on Dependencies
+
+Add-ons can depend on other add-ons and that dependency may be soft or hard. Hard dependency implies that add-on provisioning must fail if the dependency is not available. 
+For example, if an add-on requires access to AWS Secrets Manager for a secret containing a license key, credentials or other sensitive information, it can declare dependency on the CSI Secret Store Driver. 
+
+Dependency management for direct hard dependency are implemented using a decorator `@dependable`.
+
+Example:
+
+```typescript
+import { Construct } from "@aws-cdk/core";
+import { ClusterInfo } from "../../spi";
+import { dependable } from "../../utils";
+import { HelmAddOn, HelmAddOnUserProps } from "../helm-addon";
+
+export class MyProductAddOn extends HelmAddOn {
```

Comment:
we should introduce `HelmAddOn` right after `ClusterAddOn`.

Reply:
Moved helm and non-helm add-on sections up.